File: Practical_six_sem/AI_Lab/AStar8Puzzle.java
import java.util.*;

class Node {
    int[][] board;
    int g, h, f;
    Node parent;

    public Node(int[][] board, int g, int h, Node parent) {
        this.board = board;
        this.g = g;
        this.h = h;
        this.f = g + h;
        this.parent = parent;
    }
}

public class AStar8Puzzle {

    static int[][] goal = {
            {1, 2, 3},
            {4, 5, 6},
            {7, 8, 0}
    };

    // Heuristic: Number of misplaced tiles
    static int calculateH(int[][] board) {
        int count = 0;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (board[i][j] != 0 && board[i][j] != goal[i][j]) {
                    count++;
                }
            }
        }
        return count;
    }

    // Check goal state
    static boolean isGoal(int[][] board) {
        return Arrays.deepEquals(board, goal);
    }

    // Copy board
    static int[][] copy(int[][] board) {
        int[][] newBoard = new int[3][3];
        for (int i = 0; i < 3; i++)
            newBoard[i] = board[i].clone();
        return newBoard;
    }

    // Print board
    static void printBoard(int[][] board) {
        for (int[] row : board) {
            for (int val : row)
                System.out.print(val + " ");
            System.out.println();
        }
        System.out.println();
    }

    // Generate neighbors
    static List<Node> getNeighbors(Node node) {
        List<Node> neighbors = new ArrayList<>();
        int x = 0, y = 0;

        // Find blank (0)
        for (int i = 0; i < 3; i++)
            for (int j = 0; j < 3; j++)
                if (node.board[i][j] == 0) {
                    x = i;
                    y = j;
                }

        int[][] moves = {{-1,0},{1,0},{0,-1},{0,1}};

        for (int[] move : moves) {
            int nx = x + move[0];
            int ny = y + move[1];

            if (nx >= 0 && ny >= 0 && nx < 3 && ny < 3) {
                int[][] newBoard = copy(node.board);

                // Swap
                newBoard[x][y] = newBoard[nx][ny];
                newBoard[nx][ny] = 0;

                int h = calculateH(newBoard);
                neighbors.add(new Node(newBoard, node.g + 1, h, node));
            }
        }
        return neighbors;
    }

    // A* Algorithm
    static void solve(int[][] start) {

        PriorityQueue<Node> open = new PriorityQueue<>(Comparator.comparingInt(n -> n.f));
        Set<String> closed = new HashSet<>();

        Node startNode = new Node(start, 0, calculateH(start), null);
        open.add(startNode);

        while (!open.isEmpty()) {
            Node current = open.poll();

            printBoard(current.board);

            if (isGoal(current.board)) {
                System.out.println("Goal Reached!");
                return;
            }

            closed.add(Arrays.deepToString(current.board));

            for (Node neighbor : getNeighbors(current)) {
                if (!closed.contains(Arrays.deepToString(neighbor.board))) {
                    open.add(neighbor);
                }
            }
        }

        System.out.println("No solution found.");
    }

    public static void main(String[] args) {

        int[][] start = {
                {1, 2, 3},
                {0, 4, 6},
                {7, 5, 8}
        };

        solve(start);
    }
}

File: Practical_six_sem/AI_Lab/AStar8Puzzle.py
import heapq

class Node:
    def __init__(self, board, g, h, parent):
        self.board = board
        self.g = g
        self.h = h
        self.f = g + h
        self.parent = parent

    def __lt__(self, other):
        return self.f < other.f


# Goal state
goal = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 0]
]


# Heuristic: misplaced tiles
def calculate_h(board):
    count = 0
    for i in range(3):
        for j in range(3):
            if board[i][j] != 0 and board[i][j] != goal[i][j]:
                count += 1
    return count


# Check goal
def is_goal(board):
    return board == goal


# Copy board
def copy(board):
    return [row[:] for row in board]


# Print board
def print_board(board):
    for row in board:
        print(*row)
    print()


# Generate neighbors
def get_neighbors(node):
    neighbors = []
    x, y = 0, 0

    # Find blank (0)
    for i in range(3):
        for j in range(3):
            if node.board[i][j] == 0:
                x, y = i, j

    moves = [(-1,0), (1,0), (0,-1), (0,1)]

    for dx, dy in moves:
        nx, ny = x + dx, y + dy

        if 0 <= nx < 3 and 0 <= ny < 3:
            new_board = copy(node.board)

            # Swap
            new_board[x][y], new_board[nx][ny] = new_board[nx][ny], 0

            h = calculate_h(new_board)
            neighbors.append(Node(new_board, node.g + 1, h, node))

    return neighbors


# A* Algorithm
def solve(start):
    open_list = []
    closed = set()

    start_node = Node(start, 0, calculate_h(start), None)
    heapq.heappush(open_list, start_node)

    while open_list:
        current = heapq.heappop(open_list)

        print_board(current.board)

        if is_goal(current.board):
            print("Goal Reached!")
            return

        closed.add(str(current.board))

        for neighbor in get_neighbors(current):
            if str(neighbor.board) not in closed:
                heapq.heappush(open_list, neighbor)

    print("No solution found.")


# Main
if __name__ == "__main__":
    start = [
        [1, 2, 3],
        [0, 4, 6],
        [7, 5, 8]
    ]

    solve(start)